### PR TITLE
Autotest: refactoring

### PR DESCRIPTION
**Pull request by khancyr**
Prepartion work for a new autotest based on https://github.com/ArduPilot/ardupilot/pull/6284
This PR make more common functions, rework some style, change the use of print to progress function that tag the msg as AUTOTEST.

Nevertheless, until the second part is merge, it got a regression on arming and disarm on copter: the old version use RC arm/disarm, the new use mavlink message. It will be corrected on part 2 

@peterbarker as discuss this morning, I rework that so it will be simplier to integrate the second part !

### CODE REVIEW

**Comment by khancyr** [comment]
rebased on master

**Comment by peterbarker** [comment]
I'm happy enough with this.  @OXINARF - any issues if I put this in?

File: Tools/autotest/common.py [hunk 1]
```diff
@@ -45,6 +48,125 @@ def expect_callback(e):
...
+    mavproxy.send('wp save %s\n' % filename)
+    mavproxy.expect('Saved ([0-9]+) waypoints')
+    num_wp = int(mavproxy.match.group(1))
+    print("num_wp: %d" % num_wp)
+    return num_wp
+
+
+def setup_rc(mavproxy):
+    """Setup RC override control."""
+    for chan in range(1, 9):
```
Comment (khancyr) [inline Tools/autotest/common.py]: Right it should be 16 ... and the message changed ...

File: Tools/autotest/arducopter.py [hunk 2]
```diff
@@ -1072,8 +1029,7 @@ def fly_ArduCopter(binary, viewerip=None, use_map=False, valgrind=False, gdb=Fal
         print("########## Test Failsafe ##########")
         print("#")
         if not fly_throttle_failsafe(mavproxy, mav):
-            failed_test_msg = "fly_throttle_failsafe failed"
-            print(failed_test_msg)
+            print("Throttle failsafe failed")
```
Comment (khancyr) [inline Tools/autotest/arducopter.py]: yep, there is a mistmatch here ...

File: Tools/autotest/arducopter.py [hunk 3]
```diff
@@ -1282,8 +1238,8 @@ def fly_ArduCopter(binary, viewerip=None, use_map=False, valgrind=False, gdb=Fal
             print(failed_test_msg)
             failed = True
 
-    except pexpect.TIMEOUT as failed_test_msg:
-        failed_test_msg = "Timeout"
+    except pexpect.TIMEOUT as e:
+        print("Failed with timeout")
```
Comment (khancyr) [inline Tools/autotest/arducopter.py]: here too

File: Tools/autotest/common.py [hunk 4]
```diff
@@ -45,6 +48,125 @@ def expect_callback(e):
...
+        mavproxy.send('rc %u 1500\n' % chan)
+
+
+def set_switch_default(mavproxy):
+    """Set the flight mode switch to default value"""
+    mavproxy.send('rc 8 1800\n')
```
Comment (khancyr) [inline Tools/autotest/common.py]: My bad ... on copter the switch is on RC 6 ...

**Comment by khancyr** [comment]
@OXINARF The name setup_rc is not good. It should be more set_trim_all or something like that. I fact the real setup_rc could be common to copter and plane but override by rover as only rover need throttle at 1500 at start ! 

About the wait_heartbeat , it came from a race condition as SITL is at speedup 10. It happend that mavproxy doesn't catch the mode change and get stuck in infinite wait whereas the switch was toggle... What is weird is I cannot reproduce it today ...

**Comment by khancyr** [comment]
rebased and correct mistakes

File: Tools/autotest/common.py (hunk 4 above)
Comment (OXINARF) [inline Tools/autotest/common.py]: It is actually RC 5

**Comment by OXINARF** [comment]
@khancyr 

> The name setup_rc is not good. It should be more set_trim_all or something like that. I fact the real setup_rc could be common to copter and plane but override by rover as only rover need throttle at 1500 at start !

OK, so *set_rc_default* looks better now, but still there are problems:

- Plane isn't setting throttle to 1000 like it was before
- CopterAVC isn't setting throttle to 1000 like it was before

There may be more, but I've seen these two.

Also, *set_switch_default* still has the comment saying it sets the flight mode, but that's not true for Copter. Not only is the name bad, I still can't understand how we can make it common when it isn't the same for all vehicles. It is just one line, I don't see any gain from having it in common.

> About the wait_heartbeat , it came from a race condition as SITL is at speedup 10. It happend that mavproxy doesn't catch the mode change and get stuck in infinite wait whereas the switch was toggle... What is weird is I cannot reproduce it today ...

Well, that just means to me that this is a workaround instead of fixing the correct place.

@peterbarker You are much more into autotest than me, so I'll defer to you if it should be merged or not.

**Comment by khancyr** [comment]
rebased, revert set_switch_default and throttle_default
add quadplane and ardusub

**Comment by khancyr** [comment]
ping @peterbarker

**Comment by peterbarker** [comment]
```
LAND> Set RC switch override to 6 (PWM=1815 channel=5)
waypoint 0
STABILIZE> AUTOTEST: Got mode STABILIZE
rc 3 1000
arm throttle
rc 3 1000
STABILIZE> arm throttle
STABILIZE> APM: Arm: Throttle too high
Got MAVLink msg: COMMAND_ACK {command : 400, result : 4}
Mode STABILIZE
Flight battery 50 percent
```

That's a problem.   Does this get stuck for you?

Also, it gets stuck there forever, as there's no timeout in `mav.motors_armed_wait`.

**Comment by khancyr** [comment]
@peterbarker that the kind a desync I was speaking about the other day ! It appear on my computer too from time to time, but not on travis (in fact, this weekend I saw autotest failed because of that). I think it only appear when using speedup 
The problem is that we should have this situation as if you look at the log mavproxy send the RC 3 1000 ... so it should arm.

Yep motors_armed_wait doesn't have timeout ... I mostly didn't rework function, just make thing similar for to permit the transition more easily on the next PR

**Comment by khancyr** [comment]
@peterbarker ping ! I robustified the rc send command so it should pass unless it timeout on switch command ... I think it is ready to go. Solving the mavproxy timing sync should be the next task

**Comment by khancyr** [comment]
rebased and correct headers, comments, and minor fixs

**Comment by peterbarker** [comment]
I've pulled in a couple of the changes from this changeset.

Please rebase it.

Also, the set_rc function you added contains a tautology...

Please group the commits as I did for the changes I've merged; `Tools: autotest: add and use set_rc function`, `Tools: autotest: add and use arm_vehicle function`, that sort of thing.  Breaking these out into vehicle-specific commits isn't required for a final commit, and unfortunately the interleaving of commits in here means it's hard to squash the commits.

Ta!

**Comment by khancyr** [comment]
@peterbarker It should be better now !

**Comment by peterbarker** [comment]
Merged, thanks.

I left out the shebanh line for common.py; it shouldn't be run, really...

**Comment by khancyr** [comment]
very nice ! thanks !let's go to better autotest !

**Comment by peterbarker** [comment]
On Sat, 2 Dec 2017, Pierre Kancir wrote:

> very nice ! thanks !let's go to better autotest !

Speaking of which - what are your thoughts on this?

https://github.com/ArduPilot/ardupilot/pull/7343

Because I'm kinda sick of ", mav, mavproxy" :-)

**Comment by khancyr** [comment]
that looks like what I did in : https://github.com/ArduPilot/ardupilot/pull/6284 ^^ that is the next step : a common base class, and vehicle derived

**Pull request closed without merging**